File: 0410-split-array-largest-sum/0410-split-array-largest-sum.cpp
class Solution {
public:
    int getSubarrays(vector<int>& nums, int k, int maxsum){
        int count = 1;
        int sum = 0;
        for(int i = 0; i < nums.size(); i++){
            if(nums[i] + sum <= maxsum){
                sum += nums[i];
            }else{
                count++;
                sum = nums[i];
            }
        }
        return count;
    }

    int splitArray(vector<int>& nums, int k) {
        int low = *max_element(nums.begin(), nums.end());
        int high = accumulate(nums.begin(), nums.end(),0);

        while(low <= high){
            int mid = (low+high)/2;
            int subarrays = getSubarrays(nums, k, mid);

            if(subarrays <= k){
                high = mid - 1;
            }else{
                low = mid + 1;
            }
        }
        return low;
    }
};

File: 0410-split-array-largest-sum/0410-split-array-largest-sum.py
class Solution:
    def isPossible(self, nums, sub_size):
        count = 1
        summ = 0

        for i in range(len(nums)):
            if summ + nums[i] <= sub_size:
                summ += nums[i]
            else:
                count += 1
                summ = nums[i]

        return count

    def splitArray(self, nums: List[int], k: int) -> int:
        low = max(nums)
        high = sum(nums)
        ans = -1
        n = len(nums)

        while low <= high:
            mid = (low + high) // 2

            if self.isPossible(nums, mid) > k:
                low = mid + 1
            else:
                high = mid - 1
        
        return low